File: impl/php/src/Puncturable.php
<?php
declare(strict_types=1);

namespace Nyx;

/**
 * Punktierbare PRF (Whitepaper 7.7, Schicht 2).
 *
 * Byteweise identisch zu nyx/puncturable.py: derselbe GGM-Baum, dieselben
 * Ableitungslabel, dieselbe Tiefe. Ein Knoten kann dadurch von der einen
 * auf die andere Implementierung wechseln, ohne seinen Bestand zu verlieren.
 *
 * Nach der Herausgabe eines Teils entfernt der Knoten aus seinem eigenen
 * Schluessel die Faehigkeit, genau diesen Tag abzuleiten. Wird er danach
 * beschlagnahmt, kommt er an das bei ihm liegende Chiffrat nicht mehr heran.
 */
final class Puncturable
{
    public const DEPTH = 48;
    private const LEFT = "\x00";
    private const RIGHT = "\x01";
    private const LEAF = "\x02";

    /** @var array<int, array{prefix: int, len: int, seed: string}> */
    private array $cover;
    private int $punctured = 0;

    public function __construct(string $master)
    {
        $this->cover = [['prefix' => 0, 'len' => 0, 'seed' => $master]];
    }

    /**
     * Blattnummer eines Tags: die oberen DEPTH Bits des Hashwerts.
     *
     * Bei DEPTH = 48 sind das genau die ersten sechs Bytes. Der Wert passt
     * damit in einen vorzeichenbehafteten 64-Bit-Integer, und die
     * Umrechnung kommt ohne bcmath aus — die Erweiterung ist auf vielen
     * PHP-Installationen nicht vorhanden.
     */
    public static function leafIndex(string $tagRaw): int
    {
        if (self::DEPTH !== 48) {
            throw new \LogicException('leafIndex ist auf DEPTH = 48 ausgelegt');
        }
        $digest = Canonical::sha256('nyx/v1/punct-index', $tagRaw);
        $high = unpack('N', substr($digest, 0, 4))[1];
        $low = unpack('n', substr($digest, 4, 2))[1];
        return $high * 65536 + $low;
    }

    private static function bitAt(int $index, int $level): int
    {
        return ($index >> (self::DEPTH - 1 - $level)) & 1;
    }

    private static function child(string $seed, int $bit): string
    {
        return Canonical::hmac($seed, $bit ? self::RIGHT : self::LEFT);
    }

    private static function prefixOf(int $index, int $len): int
    {
        return $len === 0 ? 0 : $index >> (self::DEPTH - $len);
    }

    private function covering(int $index): ?int
    {
        foreach ($this->cover as $i => $entry) {
            if ($entry['len'] === 0 || self::prefixOf($index, $entry['len']) === $entry['prefix']) {
                return $i;
            }
        }
        return null;
    }

    public function derive(string $tagRaw): ?string
    {
        $index = self::leafIndex($tagRaw);
        $slot = $this->covering($index);
        if ($slot === null) {
            return null;
        }
        $seed = $this->cover[$slot]['seed'];
        for ($level = $this->cover[$slot]['len']; $level < self::DEPTH; $level++) {
            $seed = self::child($seed, self::bitAt($index, $level));
        }
        return Canonical::hmac($seed, self::LEAF . $tagRaw);
    }

    public function puncture(string $tagRaw): bool
    {
        $index = self::leafIndex($tagRaw);
        $slot = $this->covering($index);
        if ($slot === null) {
            return false;
        }
        $entry = $this->cover[$slot];
        unset($this->cover[$slot]);
        $this->cover = array_values($this->cover);

        $seed = $entry['seed'];
        $prefix = $entry['prefix'];
        for ($level = $entry['len']; $level < self::DEPTH; $level++) {
            $bit = self::bitAt($index, $level);
            $this->cover[] = [
                'prefix' => ($prefix << 1) | (1 - $bit),
                'len' => $level + 1,
                'seed' => self::child($seed, 1 - $bit),
            ];
            $seed = self::child($seed, $bit);
            $prefix = ($prefix << 1) | $bit;
        }
        // Das Blattsaatgut wird nicht aufbewahrt — das ist die Punktierung.
        $this->punctured++;
        return true;
    }

    public function puncturedCount(): int
    {
        return $this->punctured;
    }

    public function toArray(): array
    {
        return [
            'punctured' => $this->punctured,
            'cover' => array_map(
                fn(array $e) => ['prefix' => $e['prefix'], 'len' => $e['len'],
                                 'seed' => bin2hex($e['seed'])],
                $this->cover
            ),
        ];
    }

    public static function fromArray(array $data): self
    {
        $key = new self(random_bytes(32));
        $key->punctured = $data['punctured'] ?? 0;
        $key->cover = array_map(
            fn(array $e) => ['prefix' => (int) $e['prefix'], 'len' => (int) $e['len'],
                             'seed' => (string) hex2bin($e['seed'])],
            $data['cover'] ?? []
        );
        return $key;
    }
}

File: impl/python/nyx/puncturable.py
"""Punktierbare PRF (Whitepaper 7.7, Schicht 2).

Der Speicherknoten legt jeden Teil unter einem Schluessel ab, den er aus
seinem eigenen Speicherschluessel ableitet. Nach der Herausgabe entfernt er
aus diesem Schluessel die Faehigkeit, genau diesen Tag abzuleiten — und
behaelt alle uebrigen. Wird der Knoten danach beschlagnahmt, kann er den
herausgegebenen Teil selbst dann nicht mehr entschluesseln, wenn das
Chiffrat noch auf seinem Datentraeger liegt und der Angreifer den
vollstaendigen Speicherinhalt samt Schluesselmaterial hat.

Der Aufbau ist der uebliche GGM-Baum: aus einem Saatgut werden zwei
Kindsaatgueter abgeleitet, die Ableitung ist eine Einbahnstrasse. Der
Knoten haelt eine Menge von Teilbaum-Saatguetern, die zusammen alle
Blaetter ausser den punktierten abdecken. Punktieren heisst: den Pfad zum
Blatt aufklappen, die Geschwister behalten, das Blatt fallen lassen.

Kosten: pro Punktierung waechst die Schluesselmenge um hoechstens ``depth``
Eintraege. Das ist der Preis dafuer, dass Vergessen hier nicht auf einem
Versprechen beruht, sondern auf einer Einwegfunktion.
"""

from __future__ import annotations

from .primitives.kdf import hmac_sha256, sha256

DEPTH = 48
LEFT = b"\x00"
RIGHT = b"\x01"
LEAF = b"\x02"


def leaf_index(tag: bytes, depth: int = DEPTH) -> int:
    """Bildet einen Tag auf ein Blatt des Baums ab."""
    digest = sha256(b"nyx/v1/punct-index", tag)
    return int.from_bytes(digest[:8], "big") >> (64 - depth)


def _child(seed: bytes, bit: int) -> bytes:
    return hmac_sha256(seed, RIGHT if bit else LEFT)


def _descend(seed: bytes, index: int, from_depth: int, to_depth: int) -> bytes:
    for level in range(from_depth, to_depth):
        bit = (index >> (DEPTH - 1 - level)) & 1
        seed = _child(seed, bit)
    return seed


class PuncturableKey:
    """Speicherschluessel eines Knotens, aus dem einzelne Tags entfernbar sind."""

    def __init__(self, master: bytes, depth: int = DEPTH):
        self.depth = depth
        # (prefix_bits, prefix_len, seed) — deckt alle Blaetter unter dem Prefix ab
        self._cover: list[tuple[int, int, bytes]] = [(0, 0, master)]
        self._punctured = 0

    # -- Ableiten --------------------------------------------------------

    def _covering(self, index: int) -> tuple[int, int, bytes] | None:
        for prefix, plen, seed in self._cover:
            if plen == 0 or (index >> (self.depth - plen)) == prefix:
                return prefix, plen, seed
        return None

    def derive(self, tag: bytes) -> bytes | None:
        """Schluessel fuer einen Tag, oder None wenn bereits punktiert."""
        index = leaf_index(tag, self.depth)
        entry = self._covering(index)
        if entry is None:
            return None
        _, plen, seed = entry
        leaf = _descend(seed, index, plen, self.depth)
        return hmac_sha256(leaf, LEAF + tag)

    # -- Vergessen -------------------------------------------------------

    def puncture(self, tag: bytes) -> bool:
        """Entfernt genau diesen Tag aus dem Schluessel. Unumkehrbar."""
        index = leaf_index(tag, self.depth)
        entry = self._covering(index)
        if entry is None:
            return False

        prefix, plen, seed = entry
        self._cover.remove(entry)

        # Pfad zum Blatt aufklappen, jeweils den Geschwisterteilbaum behalten.
        node_prefix, node_seed = prefix, seed
        for level in range(plen, self.depth):
            bit = (index >> (self.depth - 1 - level)) & 1
            sibling_seed = _child(node_seed, 1 - bit)
            sibling_prefix = (node_prefix << 1) | (1 - bit)
            self._cover.append((sibling_prefix, level + 1, sibling_seed))
            node_seed = _child(node_seed, bit)
            node_prefix = (node_prefix << 1) | bit
        # node_seed waere das Blatt — es wird nicht aufbewahrt.
        self._punctured += 1
        return True

    @property
    def punctured_count(self) -> int:
        return self._punctured

    @property
    def key_size(self) -> int:
        """Anzahl der aufbewahrten Teilbaum-Saatgueter."""
        return len(self._cover)

    def __repr__(self) -> str:
        return (f"<PuncturableKey tiefe={self.depth} "
                f"punktiert={self._punctured} teilbaeume={len(self._cover)}>")
